kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt

```diff
@@ -376,6 +366,12 @@ class TunnelService : VpnService() {
             setSession(SESSION_NAME)
             setMtu(MTU)
         }.establish()!!.let {
+            // For Android we consider that we need to get the DNS to be up and running
```

Comment:
Wouldn't a better place for this be inside the `onSetInterfaceConfig` callback?

Also, it might be good to check for `tunnelState = State.DOWN` before calling `updateStatusNotification` -- I'm not sure if it's a good idea to call that repeatedly.

Lastly, if we're deciding that DNS servers must be set for the tunnel to be "Up", does that mean if they're removed, we consider the tunnel to be "Down"? And we should disconnect connlib at that point?

That seems a little hairy...

Maybe we want to tie `tunnelState` to the interface IP being set, and the `updateStatusNotification` to the DNS servers instead.

I know our big customer on Android is using DNS, but I'm still questioning whether it makes sense to tie DNS functionality to tunnel state.

Is it possible we get a linkProperties update that has no DNS servers, but DNS is in fact active and working on the system? Some browsers also use their own DNS servers (as we found out)

Reply:
FWIW, on Apple, I fixed the DNS update delay, so I consider the tunnel "Up" pretty much right after calling connect:


```swift
// Grab a session pointer
let session =
  try WrappedSession.connect(
    apiURL,
    token,
    DeviceMetadata.getOrCreateFirezoneId(logger: self.logger),
    DeviceMetadata.getDeviceName(),
    DeviceMetadata.getOSVersion(),
    connlibLogFolderPath,
    logFilter,
    callbackHandler
  )
// Attempt to set DNS right away
if let jsonResolvers = try? String(
  decoding: JSONEncoder().encode(getSystemDefaultResolvers()), as: UTF8.self
).intoRustString() {
  session.setDns(jsonResolvers)
}
// Update our internal state
self.state = .tunnelStarted(session: session)

// Start listening for network change events. The first few will be our
// tunnel interface coming up, but that's ok -- it will trigger a `set_dns`
// connlib.
beginPathMonitoring()

// Tell the system the tunnel is up, moving the tunnelManager status to
// `connected`.
completionHandler(nil)
```

Reply:
> Wouldn't a better place for this be inside the onSetInterfaceConfig callback?

Yeah, I missed that it's used for `onAddRoutes`

> FWIW, on Apple, I fixed the DNS update delay, so I consider the tunnel "Up" pretty much right after calling connect:

In android you need to consider that we don't call `setDns` until we get a callback we don't do when we call `connect`.

> Maybe we want to tie tunnelState to the interface IP being set, and the updateStatusNotification to the DNS servers instead.

Separating those doesn't seem to make sense, we want to tell the user when we consider that the tunnel is up

> Is it possible we get a linkProperties update that has no DNS servers, but DNS is in fact active and working on the system? Some browsers also use their own DNS servers (as we found out)

I think it can't happen since we only have 1 active interface

Reply:
We could use a flag to fix this, however, if dns is empty `onSetInterfaceConfig` won't be called again

Reply:
Ah, I think we could make it easier, 1 minute I'll push a commit